### DISCUSSION TITLE
How to work with loadable atoms in custom hook?

### DISCUSSION THREAD

**Discussion by richardkyk** [pmndrs/jotai#1082] [Q&A] [answered]
So I have a settings page for each program, and when the page is first visited, I set the programId on mount.
```javascript
// pages/[programId]/settings
const programId = router.query.programId ?? "";
const setProgramId = useSetAtom(programIdAtom);
const { teamMembers, teamMembersLoading, addTeamMember, deleteTeamMember } = useTeamMembers();
useEffect(() => {
  setProgramId(programId as string);
}, []);
```

I was following the discussion here [https://github.com/pmndrs/jotai/discussions/1061](https://github.com/pmndrs/jotai/discussions/1061) to arrive at the following, so I don't completely understand it but it seems to work.
```javascript
// atoms/teamMember.ts
const baseAtom = atomWithDefault(async (get) => {
  const res = await fetch(`/api/team?teamId=${get(programIdAtom)}`);
  const data = await res.json();
  if (!data) return [];
  return data as TeamMember[];
});
const asyncAtom = atom(
  (get) => get(baseAtom),
  (_get, set, arg: TeamMember[]) => set(baseAtom, arg)
);
const loadableAtom = loadable(asyncAtom);
export const finalAtom = atom(
  (get) => get(loadableAtom),
  (get, set, arg: (state: TeamMember[]) => TeamMember[]) => {
    const val = typeof arg === "function" ? arg(get(asyncAtom)) : arg;
    set(asyncAtom, val);
  }
);
```

Then I have a custom hook, (not sure if this is an anti-pattern, since invoking custom hooks creates a new instance each time, but if it is referencing the same atom it should share the same state globally is my understanding)
```javascript
// hooks/useTeamMembers.ts
const useTeamMembers = () => {
  const [res, setTeamMember] = useAtom(finalAtom);

  const addTeamMember = useCallback((member: TeamMember) => {
    setTeamMember((state) => [...state, member]);
  }, []);

  const deleteTeamMember = useCallback((userId: string) => {
    setTeamMember((state) => state.filter((x) => x.user.id !== userId));
  }, []);

  return {
    teamMembers: res.state === "hasData" ? res.data : [],
    teamMembersLoading: res.state === "loading",
    addTeamMember,
    deleteTeamMember,
  };
};

export default useTeamMembers;
```

So everything seems to work so far, but one problem is the code duplication in regards to the atom. Say if I wanted to create another hook `useProjects` where each program has x number of projects, I would need to duplicate the `atoms/teamMember.ts` but the only thing that would change is the api endpoint.

Is there a better way make the atoms reusable? I have attempted the following:
```javascript
export const atomWithAsync = <T>(url: string) => {
  const baseAtom = atomWithDefault(async () => {
    const res = await fetch(url);
    return await res.json();
  });
  const asyncAtom = atom(
    (get) => get(baseAtom),
    (_get, set, arg: T) => set(baseAtom, arg)
  );
  const loadableAtom = loadable(asyncAtom);
  const finalAtom = atom(
    (get) => get(loadableAtom),
    (get, set, arg: (state: T) => T) => {
      const val = typeof arg === "function" ? arg(get(asyncAtom)) : arg;
      set(asyncAtom, val);
    }
  );

  return finalAtom;
};

export const teamAtom = atom((get) => get(atomWithAsync(`/api/team?teamId=${get(programIdAtom)}`)));
export const projectAtom = atom((get) => get(atomWithAsync(`/api/project?programId=${get(programIdAtom)}`)));
```
however it just runs in an infinite loop. I'm not really sure how to get the `teamAtom` to depend on the `programId`, but then also make it settable once it has been initialized.

Essentially, is there a better way to have data come in from an api and initialize the state once, then allowing mutations of the atoms through custom hooks?

**Comment by dai-shi** [comment, marked as answer]
While I don't completely understand your use case yet, what you can try is probably two options: a) create atom creator function, and b) use `Provider`.

For a), I guess you can do like this?
```js
export const teamAtom = atomWithAsync('/api/team?teamId='));
```

It would be more helpful to give some ideas, if you were able to create smaller example to describe smaller problem.

**Reply by richardkyk** [reply]
Thank you for the quick response! Whilst I was setting up a codesandbox [https://codesandbox.io/s/jotai-demo-forked-rkrf1m?file=/src/App.js](https://codesandbox.io/s/jotai-demo-forked-rkrf1m?file=/src/App.js) I noticed a different issue.

Page initially loads comments for `postId = 1`
- Increment post id
- Add/delete comment
- Increment/decrement post id

It doesn't seem to be grabbing the comments for the next `postId` when the `addComment` or `deleteComment` function from the custom hook is called.

**Reply by dai-shi** [reply]
My guess with a quick look is that there's a misunderstanding with `atomWithDefault`. Once you change its value, it will lose tracking of the "default" value. So, you need to RESET it when you change the post id. It feels like `atomWithDefault` is not the right choice for your use case.

**Reply by richardkyk** [reply]
Thank you, that was helpful. I think I might have to try another approach
